kafka/src/main/java/org/togglz/kafka/KafkaStateRepository.java

```diff
@@ -0,0 +1,495 @@
+
+      kafkaConsumer.assign(topicPartitions);
+      kafkaConsumer.seekToBeginning(topicPartitions);
+    }
+
+    private List<TopicPartition> getTopicPartitions() {
+      List<TopicPartition> topicPartitions = new ArrayList<>();
+
+      for (PartitionInfo partitionInfo : kafkaConsumer.partitionsFor(inboundTopic)) {
+        topicPartitions.add(new TopicPartition(partitionInfo.topic(), partitionInfo.partition()));
+      }
+
+      return topicPartitions;
+    }
+
+    private void processRecords(ConsumerRecords<String, String> records) {
+      for (ConsumerRecord<String, String> record : records) {
+        String featureName = record.key();
+        String featureStateAsString = record.value();
+
+        try {
+          LOG.info("Starting to process state of feature {}.", featureName);
+          FeatureStateStorageWrapper storageWrapper = deserialize(featureStateAsString);
+          LOG.info("Successfully deserialized state of feature {}.", featureName);
+          updateHandler.accept(featureName, storageWrapper);
+          LOG.info("Successfully processed state of feature {}.", featureName);
+          updatePartitionOffset(record.partition(), record.offset());
+        } catch (Exception e) {
+          throw new RuntimeException("An error occurred while processing state of feature " + featureName + ".", e);
+        }
+      }
+    }
+
+    private FeatureStateStorageWrapper deserialize(String featureStateAsString) {
+      return GSON.fromJson(featureStateAsString, FeatureStateStorageWrapper.class);
+    }
+
+    private void updatePartitionOffset(int partition, long offset) {
+      LOG.info("Starting to update offset {} of partition {}.", offset, partition);
+      offsets.put(new TopicPartition(inboundTopic, partition), offset);
+      LOG.info("Successfully updated offset {} of partition {}.", offset, partition);
+    }
+
+    private void updateConsumerLag() {
+      consumerLag = accumulateEndOffsets();
+
+      if (consumerLag < 1) {
+        initializationLatch.countDown();
+      }
+    }
+
+    private long accumulateEndOffsets() {
+      AtomicLong accumulator = new AtomicLong(0L);
+
+      kafkaConsumer.endOffsets(kafkaConsumer.assignment()).forEach((topicPartition, endOffset) -> {
+        if (endOffset > 0L) {
+          long oldValue = accumulator.get();
+          long partitionOffset = endOffset - offsets.getOrDefault(topicPartition, 0L) - 1L;
+          accumulator.set(oldValue + partitionOffset);
+        }
+      });
+
+      return accumulator.get();
+    }
+
+    private void shutdown() {
+      try {
+        LOG.info("Starting to close KafkaConsumer.");
+        kafkaConsumer.close();
+        LOG.info("Successfully closed KafkaConsumer.");
+      } catch (Exception e) {
+        LOG.error("An error occurred while closing KafkaConsumer!", e);
+      } finally {
+        shutdownLatch.countDown();
+      }
+    }
+  }
+
+  static class FeatureStateProducer {
```

Comment:
Make `private` including all the methods?

Reply:
Done.